src/coap_client_zephyr.c

```diff
@@ -0,0 +1,1147 @@
+#define LOG_SOCKADDR(fmt, addr)                         \
+    do {                                \
+        char buf[NET_IPV6_ADDR_LEN];                \
+                                    \
+        if (addr->sa_family == AF_INET6) {          \
+            net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr, \
+                      buf, sizeof(buf));        \
+        } else if (addr->sa_family == AF_INET) {        \
+            net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr, \
+                      buf, sizeof(buf));        \
+        }                           \
+                                    \
+        LOG_DBG(fmt, buf);                  \
+    } while (0)
+#else
+#define LOG_SOCKADDR(fmt, addr)
+#endif
+
+static int __golioth_connect(struct golioth_client* client,
+                             const char *host, const char *port) {
+    struct zsock_addrinfo hints = {
+        .ai_family = AF_UNSPEC,
+        .ai_socktype = SOCK_DGRAM,
+        .ai_protocol = IPPROTO_UDP,
+    };
+    struct zsock_addrinfo *addrs, *addr;
+    int ret;
+    int err = -ENOENT;
+
+    ret = zsock_getaddrinfo(host, port, &hints, &addrs);
+    if (ret < 0) {
+        LOG_ERR("Fail to get address (%s %s) %d", host, port, ret);
+        return -EAGAIN;
+    }
+
+    for (addr = addrs; addr != NULL; addr = addr->ai_next) {
+        LOG_SOCKADDR("Trying addr '%s'", addr->ai_addr);
+
+        err = golioth_connect_sockaddr(client, host, addr->ai_addr, addr->ai_addrlen);
+        if (!err) {
+            /* Ready to go */
+            break;
+        }
+    }
+
+    zsock_freeaddrinfo(addrs);
+
+    return err;
+}
+
+static int golioth_connect(struct golioth_client* client, const char *host,
+                           const char *port) {
+    int err;
+
+    if (client->sock >= 0) {
+        return -EALREADY;
+    }
+
+    err = __golioth_connect(client, host, port);
+    if (!err) {
+        golioth_coap_reqs_on_connect(client);
+
+        if (client->on_connect) {
+            client->on_connect(client);
+        }
+    }
+
+    return err;
+}
+
+static int client_connect(struct golioth_client* client) {
+    char uri[] = CONFIG_GOLIOTH_COAP_HOST_URI;
+    char *host = &uri[sizeof("coaps://") - 1];
+    const char *port = "5684";
+    char *colon;
+    int err;
+
+    if (strncmp(uri, "coaps://", sizeof("coaps://") - 1) != 0) {
+        return -EINVAL;
+    }
+
+    colon = strchr(host, ':');
+    if (colon) {
+        *colon = '\0';
+        port = colon + 1;
+    }
+
+    err = golioth_connect(client, host, port);
+    if (err) {
+        LOG_ERR("Failed to connect: %d", err);
+        return err;
+    }
+
+    fds[POLLFD_SOCKET].fd = client->sock;
+    fds[POLLFD_SOCKET].events = ZSOCK_POLLIN;
+
+    return 0;
+}
+
+static int golioth_close(struct golioth_client* client) {
+    int ret;
+
+    ret = __golioth_close(client->sock);
+    client->sock = -1;
+
+    return ret;
+}
+
+static int golioth_disconnect(struct golioth_client* client) {
```

Comment:
Can we combine this with the below function? It's a little confusing to have both `client_disconnect()` and `golioth_disconnect()` and `golioth_disconnect()` is only called by `client_disconnect()`.

Reply:
Nice! Feels more organized and easier to read with the simplification and reordering.